### DISCUSSION TITLE
Why User-Defined Conversion Operators from Base to Derived are not allowed on the Derived type?

### DISCUSSION THREAD

**Discussion by fitdev** [dotnet/csharplang#6234] [General]
```cs
class Base { }

class Derived<T> : Base {
  public static implicit operator Derived<T>(Base it) { ... } // <= Compiler Error CS0553: user-defined conversions to or from a base type are not allowed
}
```

It is perfectly understandable why derived-to-base would be disallowed, after all it is already part of a runtime. But why would base-to-derived be disallowed too?

Also note that it is important to define such conversions _in the derived type_, not the base, because:
- Base type may exist in another assembly or not be under control of a user at all
- Derived class may have type parameters, which the base class does not, as in the example

**Comment by svick** [comment]
> It is perfectly understandable why derived-to-base would be disallowed, after all it is already part of a runtime. But why would base-to-derived be disallowed too?

I believe it's a similar reason: casting from base to derived already has a meaning in the language as a type test. It would be really confusing if such a cast sometimes behaved differently, so the language does not allow it.

**Reply by jrmoreno1** [reply]
@svick : could you expand upon that a bit?  I was looking into this as a method of creating typed aliases (`public class D:B{}`) with the only actual addition being the implicit conversion (so it would be a 100% safe drop in replacement) and am a bit disappointed that it doesn't work.

**Reply by colejohnson66** [reply]
Currently, `(Derived)objectOfBase` has a well-defined meaning: an upcast that will throw if `objectOfBase` is not a `Derived`. Adding a user-defined conversion operator would break this expectation. There's no technical reason it can't be done, but the practical reason of DX is why you're not allowed to.

**Reply by jnm2** [reply]
Yep, <https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/classes.md#15104-conversion-operators>:

> It is not possible to directly redefine a pre-defined conversion. Thus, conversion operators are not allowed to convert from or to `object` because implicit and explicit conversions already exist between `object` and all other types. Likewise, neither the source nor the target types of a conversion can be a base type of the other, since a conversion would then already exist.

And, <https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/conversions.md#1035-explicit-reference-conversions>:

> > *Note*: Reference conversions, implicit or explicit, never change the value of the reference itself ([§8.2.1](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/types.md#821-general)), only its type; neither does it change the type or value of the object being referenced. *end note*

**Reply by jnm2** [reply]
> (so it would be a 100% safe drop in replacement)

@jrmoreno1 "Safe" is a relative term, but what you would be doing inside such an operator is returning a separate instance, where `ReferenceEquals(baseInstance, (Derived<X>)baseInstance)` is false. Any mutable state would then not be communicated between the two instances. That seems full of pitfalls.

**Reply by BlinD-HuNTeR** [reply]
> I am a bit disappointed that it doesn't work.

I hold the opinion that the language shouldn't 100% "forbid" you from doing anything if you really want it, so I'm gonna tell you some secrets.

First of all, the rules for user-defined conversions are relaxed when generics are involved, and because of that, Roslyn will actually accept pretty much any user-defined conversion that happens to exist, it just won't let you define them. But you could define any conversion you want in IL, even conversions from `object`.

Second, there is a hidden way to create user-defined conversions in C#, that ignores all rules:
```
using System.Runtime.CompilerServices;

class Base { }
class Derived<T> : Base
{
    [SpecialName]
    public static Derived<T> op_Implicit(Base it) { ... }
}
```
The point is that, behind the scenes, user-defined conversions are nothing more than static methods named `op_Implicit` or `op_Explicit`, and marked with [SpecialNameAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.specialnameattribute). So you can just reproduce that and it will work.

However, if the project defining this method is in the same solution, the IDE will not recognize it as a user-defined conversion at design-time (it will still compile though). It's better, instead, to put such methods in a separate library, and reference the compiled DLL, not the source project. That way the IDE will recognize it as an actual conversion.

**Reply by jnm2** [reply]
Does Roslyn actually call through this `op_Implicit` (at least, today)?

**Reply by BlinD-HuNTeR** [reply]
I am under the impression that you wanna take the nice things away from me 😕

Well, to answer your question, user-defined conversions, when compiled, become methods named `op_Implicit` in metadata. So if you manually write a method with that name, and mark it with `SpecialNameAttribute`, then it will be recognized as an implicit conversion **after being compiled**.

That's why I said the IDE would not recognize it, but when compiling, it would. Because when you hit compile, it will use the already compiled references, and there is no difference, at metadata level, between a "real" conversion, and one created in this way.

P.S.: It also doesn't work if you try to use the conversion on the **same project** where it is defined. It has to be defined in one project, and used in another.

**Reply by HaloFour** [reply]
Yes, you can manually emit an assembly that has a member that mimics said implicit conversion operator.  The question is whether or not the C# compiler even respects the presence of that operator in this case.  I would expect that it doesn't, that it emits a cast regardless of whether such an method exists.  But someone with more time on their hands could confirm/deny this.

**Reply by CyrusNajmabadi** [reply]
Also note that such things are in a gray area in terms of support.  They could work in one release, and fail in another.  Constructs outside of what c# supports can and do break occasionally.

**Reply by BlinD-HuNTeR** [reply]
>The question is whether or not the C# compiler even respects the presence of that operator in this case.

Yes it does, although it will not choose a user-defined conversion over a built-in conversion. But for the OP's case, this is a conversion from Base to Derived, which is by default an explicit conversion. So you can write an **implicit** conversion for that case, which is normally forbidden by the language, but if it exists, it will be used.

**Reply by jnm2** [reply]
So it will not call the `op_Implicit` method, and instead it will (implicitly) crash with InvalidCastException if the `Base` was `Derived<SomeOtherT>`?

**Reply by BlinD-HuNTeR** [reply]
> So it will not call the `op_Implicit` method, and instead it will (implicitly) crash with InvalidCastException if the `Base` was `Derived<SomeOtherT>`?

It *will* call `op_Implicit` if you do it implicitly (`Derived<T> obj = instanceOfBase`). If, however, you try to do an explicit cast (`Derived<T> obj = (Derived<T>)instanceOfBase`), then it will emit a builtin conversion, which will faill with `InvalidCastException` at runtime.

**Reply by CyrusNajmabadi** [reply]
what if you have both op_Implicit and op_Explicit?

**Reply by jnm2** [reply]
Nice, thanks! It's nice to know what it does (again- today). As you probably realize, I'm not trying to take away nice things from you per se, but pointing to likely outcomes for using it. If those two things seem equivalent to you, I'll let you sort that out. 😉

**Reply by BlinD-HuNTeR** [reply]
> what if you have both op_Implicit and op_Explicit?

The immediate answer is, that `op_Explicit` cannot win over the built-in conversion already existing in the language, so it will be ignored. But, things can get more complex when chains are involved (casting from C to A through a base class B, that could even be `object`).

However none of this is unpredictable, the actual rules are those that apply when generics are involved. Take the example:
```C#
class MyClass<T> { public static implicit operator MyClass<T>(T t) { ... } }
```
If you have a variable or property of type `MyClass<object>`, then it will be implicitly castable from `object`, which means that it's also implicitly castable from literally anything, except interfaces.

And the point is, if such implicit conversion works with generics, then it will also work if you make `MyClass` non-generic, and write the conversion as `[SpecialName] public static MyClass op_Implicit(object obj) { ... }`. This will be a non-generic class that's implicitly castable from `object`.
>I'm not trying to take away nice things from you per se, but pointing to likely outcomes for using it

No problem, I have years of experience of doing undocumented/unsupported/undefined behavior things, and I understand the implications. I acknowledge that none of this is officially supported, but I don't see any risks for this specific case.

The point is, Roslyn doesn't make any distinction between generics and non-generics, it has a well-defined algorithm to find user-defined conversions, and if the conversion exists, it will be used, regardless of being generic or not. And unless someone explicitly makes a change to make Roslyn ignore these "hacked" conversions, they will not stop working in the future.
